semgrep-core/src/fixing/Autofix_printer.ml

```diff
@@ -0,0 +1,128 @@
+(* Nat Mote
+ *
+ * Copyright (C) 2019-2022 r2c
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public License
+ * version 2.1 as published by the Free Software Foundation, with the
+ * special exception on linking described in file LICENSE.
+ *
+ * This library is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the file
+ * LICENSE for more details.
+ *)
+
+open Common
+open AST_generic
+module MV = Metavariable
+
+(******************************************************************************)
+(* Handles AST printing for the purposes of autofix.
+ *
+ * The main printing logic happens elsewhere. This module's main purpose is to
+ * extend the existing AST printers so that they can avoid printing AST nodes
+ * which are lifted unchanged from either the target file (via metavariable
+ * bindings) or the fix pattern in the rule. This serves two purposes:
+ * - It lets us synthesize autofixes without having to implement printing for
+ *   all of the AST nodes in the fixed AST.
+ * - It lets us make minimal changes to target files by carrying over
+ *   formatting, comments, etc. from the original code.
+ *)
+(******************************************************************************)
+
+module PythonPrinter = Hybrid_print.Make (struct
```

Comment:
functors and objects, wow, we're using advanced features of OCaml this week (brandon started to use private constructors).